internal/observe/read_closer.go

```diff
@@ -0,0 +1,42 @@
+package observe
+
+import (
+	"errors"
+	"io"
+)
+
+// ReadCloser is an observable io.ReadCloser.
+type ReadCloser interface {
+	io.ReadCloser
+	EOF() <-chan struct{}
```

Comment:
Does this make sense to be a chan or should it just be a bool ( or maybe even reuse the io.EOF error type )?

Reply:
If we don't use a channel we'd have to poll the reader waiting for it to be at an `EOF` state. The channel allows us to respond to the `EOF` as an event.